### ISSUE TITLE
logrotation problems in case of crash during the rotation

### ISSUE THREAD

**Issue by sponce** [xrootd/xrootd#302]
In the logrotate script of xrootd, one can find the following postrotate script : 
```bash
/var/log/xrootd/*/*.log /var/log/xrootd/*.log
...
  postrotate
     PIPE="`dirname $1`/.`basename $1`"
     if [ -p $PIPE ]; then
       /usr/bin/expect -c "set timeout 2; spawn /bin/sh -c \"echo ping > $PIPE\"; expect timeout { exit 1 } eof { exi    t 0 }" > /dev/null;
     fi
  endscript
```
This has 2 flows. 
The first one explains the title : in case you have an xrootd crash (or stop) in between the 'if [-p...]' check and the 'echo ping > $PIPE' statement, you will actually create a regular file with content 'ping' on the filesystem rather than waking up xrootd.
Not a big deal so far. One file of 4 bytes won't kill your system.

The problem will come on the next iteration, and is a generic second flow : the -p check on $PIPE is doing shell expansion. Now the value of PIPE is (on second iteration) /var/log/xrootd/.*.log
So if another file is there matching this pattern on top of the expected .xrootd.log, the result of the expansion will be several words, breaking the check and leading to : 
```
logrotate_script: line 2: [: /var/log/xrootd/.bla.log: binary operator expected
```
Note that here I did a simple `touch /var/log/xrootd/.bla.log`, but in practice, you get it from the first problem of having a .*.log file and the error looks like : 
```
logrotate_script: line 2: [: /var/log/xrootd/.*.log: binary operator expected
```

**Comment by sponce** [comment]
We seem to be able to trigger it pretty regularly which is forcing us to spend our time dropping the spurious .*.log on a daily base.
Now this statistic makes me think that it's not only bad luck that we restart xrootd in between the -p check and the echo statement. I believe that every single restart of xrootd triggers triggers log rotation somehow (or at least some equivalent code) and that we thus have a race condition during the restart.
Anyway, it would be nice to fix this as soon as possible as it makes the administration of large xrootd clusters a real pain.

**Comment by sponce** [comment]
I have some new input concerning this : it can be reproduced without xrootd being even restarted, as is proven on this machine (lxbre43a05) :
# ls -alrt /var/log/xrootd/ | grep log
prw-------.  1 root   root       0 Mar 10  2015 .xrootd.log
-rw-r--r--.  1 root   root    1325 Mar 12  2015 xrootd.log.1
-rw-r--r--.  1 root   root       5 Jan 22 03:37 .*.log
# ps auxf | grep xroo[t]
stage    198202  0.0  0.0 579880 15956 ?        Sl    2015   0:16 /usr/bin/xrootd -l /var/log/xrootd/xrootd.log -c /etc/xrd.cf.server -k fifo -b -s /var/run/xrootd/xrootd-server.pid -n server

After some digging with Michal, we've found out 3 issues : 
  - a bug in logrotate by which the given regexp is considered for log rotation if no expansion of it is existing, this explains why *.log is considered when no .log file exists. Not fixed to my knowledge
  - a bug in logrotate (https://fedorahosted.org/logrotate/ticket/8) by which the postrotate script is called for files not rotated. So now we call the postrotate for *.log. This has been fixed (https://fedorahosted.org/logrotate/changeset/403) and is available in logrotate 3.8.4. So CC7 is immune and slc6 should probably upgrade from 3.7.8, that is 7 years old !?!?!
  - a bug in the postrotate of xrootd by which the file name is expanded in the -p check while it should not be. So double quotes need to be added.

Last one can be easily done and fixes all issues.
